### ISSUE TITLE
How can we move the temrinal panel to the right?

### ISSUE THREAD

**Issue by noahehall** [microsoft/vscode#141349]
We use to be able to execute `move panel right` for moving the terminal panels to the right, however in the latest version of vscode (not sure where it started) that cmd has been deprecated, now we can only move the panel from the left side or docked on the bottom.

How can we move the panel to the right? im almost certain im doing something wrong

*main content* | *terminal panels stacked vertically* | *explorer panels stacked vertically*

the above UX is essential, as in this setup toggling the terminal/explorer does not cause the main content to jump around if the main content is flush left

**Comment by gjsjohnmurray** [comment]
pinging @sbatten (bot made an assignment which surprised me)

**Comment by sbatten** [comment]
The new panel is limited to being placed opposite the side bar. I will leave this issue open to gather feedback.

**Comment by hello-smile6** [comment]
@webdev03 Could you upvote? People deserve options.

**Comment by ArturoDent** [comment]
See https://stackoverflow.com/a/70982514/836330 for a partial "workaround" in the meantime.


![terminalSideBar](https://user-images.githubusercontent.com/31813701/152488065-3fb76314-328c-4ee9-a91b-e142585b6e81.jpg)

**Comment by gotofritz** [comment]
As noted in #142075, these arbitrary changes are baffling in that they remove functionality we have been used to for years without solving any issue faced by users. It's unclear who decided on the changes and why, it looks like changing it for the sake of it. In more details

- It is longer possible to have both the terminal and the sidebar to the right of the editor. Why? What problem is removing existing functionality supposed to solve?
- When the terminal is in vertical view, it is no longer possible to make it full screen. Again, whose decision, and what problem were they trying to solve?
- It is possible to move the "views" out of the "panel". Then CTRL-J will toggle the visibility of an empty  "panel". What's the point of that? Who needs to view an empty panel?
- Conversely, there if is no shortcut to toggle the side panel visibility. Don't you collect metrics on how we use the editor? Hiding the terminal, editing code, showing the terminal again, sometimes maximised, sometimes not, it's what we do. All. Day. These changes completely destroy that. And to what benefit?
- The abstractions are hard to understand and explain to non initiated. The wording in the menu "Move views to panels" is too crypting. "Panels"? "Views"? And "Sidebar" and "Side Panel" are too similar. As a user, why should I learn your jargon, and why should I care? The "panel" is an abstraction used by VSC developers. We users don't care about your implementation details, we care about concrete elements like the "Terminal". As you probably know, the number 1 rule in UX is "don't make me think". 
- The menu entry "Terminal" in the menubar has always been associated with the terminal in the panel, now it opens a completely separate terminal in one of the document windows. Why? Why have two separate things doing the same thing? Why have the menu control only the one and not the other? You do understand how chaotic it all gets when you have terminals in the side panels, and terminals in the editor window, and you are trying to remember what is in which window and which shortcut makes what panel appear... AAARGGH

The whole design decision is a complete train wreck. I urge you to reconsider. Thanks.

**Comment by gjsjohnmurray** [comment]
> * Conversely, there is no shortcut to toggle the side panel visibility.

@gotofritz on this particular point I assume you know you can assign whatever shortcut suits you:

![image](https://user-images.githubusercontent.com/6726799/152506286-28dfb68a-2128-4bb6-8ffd-96137856a7e3.png)

**Comment by gjsjohnmurray** [comment]
> * The menu entry "Terminal" in the menubar has always been associated with the terminal in the panel, now it opens a completely separate terminal in one of the document windows.

For me this only happens after I set `"terminal.integrated.defaultLocation": "editor"`

This capability arrived last year (see https://code.visualstudio.com/updates/v1_58#_terminals-in-the-editor-area).

**Comment by gotofritz** [comment]
> > * Conversely, there is no shortcut to toggle the side panel visibility.
> 
> @gotofritz on this particular point I assume you know you can assign whatever shortcut suits you:
> 
> ![image](https://user-images.githubusercontent.com/6726799/152506286-28dfb68a-2128-4bb6-8ffd-96137856a7e3.png)

Yes of course. I wanted to flag that changing an existing shortcut to toggle a (potentially) empty panel, while the expected functionality is left to individual users to wire up manually in the preferences, is not ideal IMHO

**Comment by gotofritz** [comment]
> > * The menu entry "Terminal" in the menubar has always been associated with the terminal in the panel, now it opens a completely separate terminal in one of the document windows.
> 
> For me this only happens after I set `"terminal.integrated.defaultLocation": "editor"`
> 
> This capability arrived last year (see https://code.visualstudio.com/updates/v1_58#_terminals-in-the-editor-area).

You are right, thanks for the tip... there are so many obscure terminal preferences settings that I may well have changed that manually in my desperate quest to bring back the old functionality

**Comment by gjsjohnmurray** [comment]
> See https://stackoverflow.com/a/70982514/836330 for a partial "workaround" in the meantime.

Putting terminals in a dedicated editor group on one side could be improved by addressing this issue:

#16927 - Allow a maximized editor group to use the full width rather than leaving 220px per group for each of the other editor groups. Also make `View: Toggle Editor Group Sizes` and friends remember the original group splitbar position when maximizing a group, reinstating that when toggling back from maximized. Currently it resets to even distribution, in the same way double-clicking on the splitbar does.

**Comment by adam-arold** [comment]
@sbatten as I've stated elsewhere this is not a missing feature. It is a broken feature thus a breaking change in `1.64.0`!

**Comment by adam-arold** [comment]
> @webdev03 Could you upvote? People deserve options.

Upvoted, although this shouldn't be up for voting. This whole thing should be rolled back.

**Comment by avianion** [comment]
This is an absolute joke, how is it possible to make such a regression in your softwares functionality! This feature has been working for years.

**Comment by gjsjohnmurray** [comment]
Apparently no-one who uses that functionality does so with Insiders. Or if they do, I haven't seen an issue raised about it ahead of 1.64 shipping as Stable yesterday.

Workaround for the moment is to revert to 1.63. Installation links are at https://code.visualstudio.com/updates/v1_63 and you may also want to turn off update checks.

I am confident that the team (of which I am not a member) will be taking this feedback seriously.

**Comment by carlfriedrich** [comment]
I'd like to add some general and specific thoughts here.

### General thoughts

In general, having an additional panel might be useful for some users. Adding such a new panel, however, should not break anyone's workflow who does NOT need another panel. In an ideal case, users who don't need it should not even recognize that it has been introduced. In a semi-ideal case, they should at least be able to re-configure their previous workflow in the new environment. Both cases are not true for the introcution of the side panel, which makes this a **breaking change** and should be announced accordingly, in my opinion.

### Missed opportunities

I assume that the developers didn't put enough effort into evaluating existing workflows and maybe didn't even notice the impacts of this change. It could have easily been designed non-breaking, e.g. as follows:
- Rename the existing "Panel" to "Primary Panel", keeping all functionality and keybindings as they are, just a change of wording.
- Add a new general "Secondary Panel" with the same functionality as the existing panel.
- Duplicate all existing panel-commands, so that there is a version for the primary and one for the secondary panel.
- Make the secondary panel be able to be placed at any location, just like the primary panel (left, right, top, bottom).

I don't see why the new side panel has been designed so different from the already existing panel. If I needed another panel, why force a specific layout? Who says I couldn't need two side panels? One at the left and one at the right? Or two bottom panels?

However, now that the new side panel has been introduced, it might be too late to design it this way, because that would make it *another* breaking change. I still think that this would be the most flexible, least surprising way, though.

### The workaround

I don't think that the proposed "workaround" by putting the terminal into an editor group on the right side can replace previous workflows. The missing possibility to maximize the terminal has already been noted here. However, there are more limitations:
- The terminal cannot be hidden. In order to hide it, one would have to close the editor group and open a new terminal, which is not the same.
- There is no command to open another terminal in the same editor group. When having the terminal in the panel, this is possible using the `Terminal: Create Terminal` command.
   - Running `Terminal: Create Terminal` while having a terminal open in an editor group, the new terminal is opened in the (side) panel.
   - Running `Terminal: Create New Terminal in Editor Area` opens the terminal in the left-most editor-group, which is not the right one.
   - Running `Terminal: Create New Terminal in Editor Area to the Side` opens the terminal in a new editor group, having the terminal side by side.
- Splitting the terminal vertically does not work anymore. In the panel this can be done using the command `Terminal: Split Terminal`. When having a terminal open in an editor group, this creates a terminal in a new editor group on the right side.
- The terminal cannot be resized using hotkeys. In the panel this is possible using the `Terminal: Resize Terminal [Up/Left/Right/Down]` commands. (As a side note, these commands do not work with the new side panel as well, which is already tracked in #142176).

### Proposed steps

Taking into account all the above, I see two options in order to make previous workflows possible again:

1. **The clean solution**

- Create a fix release (fast!) with the new side panel removed, restoring all previous functionality.
- Take the time to re-implement the side panel in the form of a secondary panel like proposed above under "Missed opportunities".

2. **The fix solution**

- Do not tie the position of the vertical panel to the position of the side bar. Make it configurable independently, so that it can be on the left or on the right.
- Add a command `View: Toggle Maximized Side Panel` analogous to `View: Toggle Maximized Panel`.
- Make the `Terminal: Resize Terminal [Up/Left/Right/Down]` commands work with the side panel again (see #142176, as already mentioned).

### The bottom line

Of course the fix solution would be the easier one, but in my opinion it is far less sustainable as it bears the risk of future insonsistencies. From a developer's perspective, having two differently designed UI objects providing essentially the same functionality just leads to a higher maintenance effort due to more corner cases.

**Comment by adam-arold** [comment]
> Apparently no-one who uses that functionality does so with Insiders. Or if they do, I haven't seen an issue raised about it ahead of 1.64 shipping as Stable yesterday.

I don't think that this is a valid criteria for introducing breaking changes. This is similar to the reasoning that was in [The Hitchhiker's Guide to the Galaxy](https://en.wikipedia.org/wiki/The_Hitchhiker%27s_Guide_to_the_Galaxy):

> Oh, you could comment on this issue it was on the Galactic Bulletin Board. (of course it didn't matter that humanity didn't know about the existence of the bulletin board nor had the tools to travel interstellar travel.

**Comment by adam-arold** [comment]
To give you a very specific example, this is what I can't do with `1.64.0`:
![vscode](https://user-images.githubusercontent.com/1253248/152531531-823a00d9-a4c1-4fd3-a6c9-dc35a113a448.gif)

**Comment by adam-arold** [comment]
Note that #141569 is **very closely related** to this issue, in fact, that bug was introduced with the same change. If you vote for this please also vote for that.

**Comment by superole** [comment]
> Workaround for the moment is to revert to 1.63. Installation links are at https://code.visualstudio.com/updates/v1_63 and you may also want to turn off update checks.
Reverting to 1.63.2 worked like a charm :) but where do I turn off automatic updates?
 
> I am confident that the team (of which I am not a member) will be taking this feedback seriously.
Me too! :D looking forward to the next smooth release (most of them have been. Really)!

_Stay positive! Work together!_

**Comment by ndevvy** [comment]
the decision to have the "side panel" and "side bar" (these names.. 🥴) limited to being opposite each other is, to put it mildly, confusing.  my workflow has pivoted for years around vertically oriented `(Side bar | Terminal | Editors )` and frequent toggling of the first two / maximizing terminal.

**Comment by cameronove** [comment]
I'm sitting here writing code (workflow = `Side bar | Terminal | Editors` same workflow I've used for years) and notice that I have an update available.  I need a break so I start update, go get coffee, come back and nearly have a heart attack.  I learned something new about myself today though - when that happened my heart literally started racing and I got short of breath - I didn't realize how important my layout was to me (I just never thought about it).

There seems to be no way to go back to `Side bar | Terminal | Editors` unless I revert to 1.63.  That's a pretty crummy thing to do to someone.

**Comment by clytras** [comment]
I have mentioned this on https://github.com/microsoft/vscode/issues/132893 comment https://github.com/microsoft/vscode/issues/132893#event-5934002677 back in December.

There I asked if we can move the new Side Bar at the Left right next to the Explorer Side Bar and that it would be great for ultra wide monitors, and I only got a single response.

![Ultra wide monitor with outline on left](https://user-images.githubusercontent.com/6103077/144689580-e902ef52-8a90-4c18-bf6f-0f8cc1eafb7a.png)

I didn't expect my proposal to be a thing just a thought, but frankly, I never expected a decision like this to get rid of the Panel "Move to the Left" as a column as it was before.

It was working and it's now broken. Please fix it by either reverting the Panel "Move to Left/Right" as a column option, or by making the new Side Bar be able to stick with the Side Panel at either side right or left.

**Comment by adam-arold** [comment]
>  [...]. I'm going to have to settle for the `editor left | terminal right` so the side-bar is on the left. This really sucks eggs.

We don't have to settle. If they don't fix this we can simply fork the codebase and use that.

**Comment by Acumane** [comment]
A note for those using WSL that're downgrading to the [previous version](https://code.visualstudio.com/updates/v1_63):
You'll also need to manually downgrade your "Remote - WSL" extension from `v0.64.2` to `v0.63.13` (or just reinstall)!

**Comment by gRizzlyGR** [comment]
If you use snaps, you can revert to revision 85 (version 1.63.2):
```bash
sudo snap revert code --revision 85
```

**Comment by adam-arold** [comment]
It is a shame that people have to revert and **no one** from the team (apart from a bot) is addressing this breaking change.

**Comment by adam-arold** [comment]
@sbatten will you address any of our concerns apart from wrongly assigning the `feature-request` tag?

**Comment by adam-arold** [comment]
> Apparently no-one who uses that functionality does so with Insiders.

Apparently, no-one at Microsoft cares about introducing breaking changes. Is there anybody who works on VS Code who is willing to address this issue?

**Comment by gtodd876** [comment]
This is infuriating. Thank you for the tip on rolling back to solve the issue now and I can get my work done. Going to look into other editor options for the future.

**Comment by dewetblomerus** [comment]
I've had my terminal on the right for years. This makes logical sense to me because all the languages I know are read left to right, I write my code on the left and I see the result of it in the terminal on the right.

I have also had my sidebar on the right for years because I work with it hidden until I need it, and when I toggle it, it only takes space from the terminal, and toggling the sidebar does not move one character of code on my screen, it only moves terminal output when I toggle it.

Now you broke all that and you are considering the fix a "feature request" on your backlog 😞

**Comment by carlfriedrich** [comment]
> I've had my terminal on the right for years. This makes logical sense to me because all the languages I know are read left to right, I write my code on the left and I see the result of it in the terminal on the right.
> 
> I have also had my sidebar on the right for years because I work with it hidden until I need it, and when I toggle it, it only takes space from the terminal, and toggling the sidebar does not move one character of code on my screen, it only moves terminal output when I toggle it.

That's exactly how I'm using it as well, for exactly the same reasons.

**Comment by sbatten** [comment]
Thanks everyone for being passionate users of VS Code and sharing your feedback. There is lots of goodness in the comments above and I hear you. When making this change, I reviewed the affected population, but I underestimated the impact on your workflows, and I am sorry about that. So, I will be pushing a fix which is simply to bring back the ability to move the, as some have called it, primary panel. The plan for this fix is to release it into the insiders build tomorrow and the recovery build for stable later this week. Going forward, we will continue to improve the new side panel while playing nicely with the primary panel in different locations. Future changes will always be in insiders first, and we hope you continue to share your ideas with us.

**Comment by gotofritz** [comment]
It took a lot of shrieking and complaining, but thanks for eventually listening to the users, @sbatten

**Comment by ArturoDent** [comment]
Testing this in Insiders today.  Put the Side Bar on the right and use the command `View: Move Panel Right`.  In one workspace with the setting `Workbench > Panel: Default Location` : `right` it put the Panel to the right of the Side Bar (so far right) but triggering the command fixed that.   Editor | Terminal (Panel) | Side Bar.  And working if put on the left too.

**Comment by marcomilone** [comment]
I usually works on 2 screens.

Left Screen:
 + Explorer(git, etc)
 + Terminal

Right Screen:
 + Editor

So Terminal should be between Explorer and Editor

I'm here to upvote to consider to improve recent change.


**Could be possible to move that SIDE panel where I want?**

Semantic talking: SIDE is a position. I will refer to it without a positional word.

Thanks for the hard work.

**Comment by hediet** [comment]
Fix landed in release/1.64

**Comment by fdiasr** [comment]
@hediet can you give a perspective when the new version will be avaialable ?  I keep seeing 1.64.1  as the lastest version.

**Comment by gjsjohnmurray** [comment]
@fdiasr see #142605. My guess is we'll have it by next week.

**Comment by ArturoDent** [comment]
@heartacker That isn't possible.  The Side Bar and Side Panel are always on opposite sides.  You would need to file another issue for that.

**Comment by adam-arold** [comment]
> I guess I'm switching to Atom now. Thanks for breaking your whole editor.

What the fluff, man. They just told us that it gets fixed in the next patch.

**Comment by ArturoDent** [comment]
I just received v1.64.2 which has this fix.

**Comment by martin-braun** [comment]
Unfortunately this fix only allows to move the bottom panel to the right, but I would also enjoy to move the new side panel to the same side like the explorer.

**Comment by gjsjohnmurray** [comment]
> It simply broke my keybinds and the most used workflow in the middle of last 3 days of release.

@h3d0 maybe not a good time to have updated VS Code, particularly if you haven't also been using Insiders alongside Stable.

**Comment by martin-braun** [comment]
@h3d0 I can see how this caused some confusion in the first place. Please take a look at #26777 to see why this was added, so things start to make sense to you hopefully. Overall it was a good change and I hope you can see it the same way.

The new side-panel is purely optional and you are able to continue to use VSCode like you did before, due to some reverting changes of 1.62.2 as you just figured out. The new side panel was never meant to be a replacement for your bottom panel that you've put to the right previously.

**Comment by adam-arold** [comment]
> I just wanted to share my expression and frustration of how bad the whole idea was. Have being using VSCdoe for 7+ years in my 12+ years developing experience. And this single change has been one of the most annoying in my whole IDE experience.

I totally agree with all your points, especially this one. I sat down one day to work on my project just to find that my workflow is **totally broken**. I've used dozens of editors throughout the years (Eclipse, Netbeans, Emacs, Idea, Vim, Notepad++, etc) and for a while I thought that VS Code was in the sweet spot. I no longer think that as **none** of those tools had a breaking change like that.

The dream is dead. And the worst part is not that when I start this I can expect this to happen but **how it was handled**. **No one** from the team acknowledged that this was a mistake, no one apologized and no one came here and said *sorry guys, we messed up, *not gonna happen again**.

I thought that Microsoft had very high standards, but apparently, they don't.

**Comment by carlfriedrich** [comment]
> **No one** from the team acknowledged that this was a mistake, no one apologized and no one came here and said _sorry guys, we messed up, _not gonna happen again__.

Well, that's not true. @sbatten [did that](https://github.com/microsoft/vscode/issues/141349#issuecomment-1033046397), and he provided a fix for the issue.

And while I can understand any anger about this and Microsoft in general (I was annoyed by the change as well): let's not forget that the developers are still humans. Repeatedly posting how annoyed you are probably won't increase anyone's motivation to address your thoughts.

**Comment by jsbrain** [comment]
> **No one** from the team acknowledged that this was a mistake, no one apologized and no one came here and said _sorry guys, we messed up, _not gonna happen again__.

That is simply not true: https://github.com/microsoft/vscode/issues/141349#issuecomment-1033046397

**Comment by clytras** [comment]
@adam-arold that's not fair. I think their response and actions were quite fast and everything is working as it was before now. And they actually did apologize to the users for this. To me, it looks like you have some generalized frustration towards Microsoft and their products. I actually thank every VSCode developer and contributor because they all made this amazing editor/IDE. And as stated above, we are humans and mistakes can happen and as long those mistakes get addressed and fixed it's fine; this is not a paid product, it's an open source project regardless if Microsoft may be having some of their developers to work on it.

Thank you @sbatten for working on this and fixing it.

**Comment by gotofritz** [comment]
Oh come on guys, it's true that @sbatten didn't cover himself in glory early on, but he learnt from his mistakes and the issue is now fixed (thanks!). Do we really need to keep going on about it? Maybe it's time to lock the thread....

**Comment by heartacker** [comment]
@gotofritz agree.

And let's look forward:

- https://github.com/microsoft/vscode/issues/142742

- https://github.com/microsoft/vscode/issues/142832

**Comment by adam-arold** [comment]
> To me, it looks like you have some generalized frustration towards Microsoft and their products.

You're right, this might be the case. I stand corrected.

**Comment by dewetblomerus** [comment]
I complained on here as loud as I felt appropriate. After the fix came in I just wanted to come back and say thank you, and now I see people still piling on.

Software lesson: After convincing the devs that it's a bug and they need to fix it RIGHT NOW, and then they fix it, you look like you don't understand software when you then steer the conversation toward the feature that caused the bug. That feature is a separate conversation, that you probably don't understand if you are still complaining in this thread.

Life lesson: When a human makes a mistake and (after some dialogue) apologizes and fixes their mistake, it is insensitive, emotionally-tone-deaf, and probably inhumane to keep going on about what a bad idea it was.

**Comment by adam-arold** [comment]
> Life lesson: When a human makes a mistake and (after some dialogue) apologizes and fixes their mistake, it is insensitive, emotionally-tone-deaf, and probably inhumane to keep going on about what a bad idea it was.

Or maybe the person in question has autism and you haven't considered that as an alternative to being insensitive, emotionally tone-deaf and inhumane.

**Comment by liborty** [comment]
How exactly is it fixed? 
I still can not get the terminal to the left side like I used to have it,
it insists on opening in the side panel on the right.
I thought I did manage somehow but then it just went back.

**Comment by dewetblomerus** [comment]
> How exactly is it fixed? I still can not get the terminal to the left side like I used to have it for the life of me. It insists on opening in the side panel on the right. There are some 80 settings to do with the terminal and not one of them will do this simple thing.
> 
> P.S. Please do not tell me off for being insensitive for daring to complain about it.

I don't believe it is a setting. I would like it to be so I can default open the terminal in the same place across all environments. And you would be forgiven for spending your efforts looking in the settings because so many other things about the UI can be changed in settings.

Right-click toward the top of the terminal next to the tabs `PROBLEMS`, `OUTPUT`, etc, then you can choose `Move Panel Right` or `Move Panel Left`. If you don't see those options try `Move Views to Panel` and then those options should appear.

**Comment by liborty** [comment]
Thanks. I was stuck in catch 22. Those options were not showing because I did not activate them with "move views from side panel to panel". On the other hand, once I finally discovered this command, I now no longer need them.
